WIP: Update redwood

bozdoz opened the pull request:
I had issues bringing up the project and I realized everything was out-of-date.  Ran `yarn rw update`, and compared with a brand new project to fix up any issues.  

Also began fixing the scrape function.  Appears to be getting the same data for each graph.   Maybe a WIP:

![image](https://user-images.githubusercontent.com/1410985/117918949-7cac5380-b2c2-11eb-81cf-d4e14eeae499.png)

![image](https://user-images.githubusercontent.com/1410985/117918970-89c94280-b2c2-11eb-96fb-90ac1b2d8df2.png)

bozdoz commented:
Resolves #12

lachlanjc commented:
Wow, thanks for bringing this back to life!! Yeah, this was one of the first releases of Redwood we ran it on. Some observations/questions:

- Looks like netlify.toml needs an update — the build logs say the `db up` command has changed to `prisma migrate`
- We implemented the `redwood-hack` because the database transactions wouldn't run without it, can we drop that now?
- Since this project is just using JS without TS, & no authentication, why are the new types files needed?
- How does the new scraping work? We just go to the URL of that function to update the database? Will it backfill to when I stopped manually updating it last summer?
- Looks like the chart data fetching isn't working in prod: https://covid19-qhsp5dim1-lachlanjc.vercel.app/

bozdoz commented:
Hmm, that's curious.

I will try the site out in docker, to make sure there's a straight-forward approach to getting this set up.  That should help figure out netlify (don't have much experience with it).

I imagine `redwood-hack` is useless, but I'll test it out in docker.

Those new types files were added automatically (and updated frequently!).  Might be worth adding them to .gitignore.

New scraping just altered the selector for the script tag, and parsed the data differently; that's it.  shouldn't be anything new.  Maybe the script tags are in a different format now, and maybe that's why the daily stats numbers are in error.  I'm not sure what it was doing before though.

lachlanjc commented:
Interesting. Yeah, let's ignore the types. The scraper never worked consistently—that was a project we started with the hopes the site could update itself, but I ended up just manually entering the numbers into the database using SQL every day until last summer, when the visualization wasn't that useful anymore.

bozdoz commented on api/prisma/migrations/migration_lock.toml:
```diff
@@ -0,0 +1,3 @@
+# Please do not edit this file manually
+# It should be added in your version-control system (i.e. Git)
+provider = "sqlite"
```
This might be an issue to commit

bozdoz replied:
Actually I think this is just so that it will only run the migration if the provider matches

bozdoz commented on api/prisma/migrations/20210512024638_brand_new/migration.sql:
```diff
@@ -0,0 +1,38 @@
+-- CreateTable
+CREATE TABLE "Day" (
```
This new migration is incompatible with production. Likely needs to be reset

bozdoz commented on api/prisma/schema.prisma:
```diff
@@ -1,36 +1,38 @@
-  newCases Int
+  id                Int     @id @default(autoincrement())
+  date              Day     @relation(fields: [dayId], references: [id])
+  country           Country @relation(fields: [countryId], references: [id])
```
This was automatically formatted and changed by prisma

bozdoz commented on api/prisma/seed.js:
```diff
@@ -50,24 +50,39 @@ const countries = [
-  }
+  },
+  // {
+  //   iso: 'cad',
```
I don't know what iso is used for

bozdoz commented on api/src/functions/graphql.js:
```diff
@@ -3,24 +3,19 @@ import {
-const services = importAll('api', 'services')
+import schemas from 'src/graphql/**/*.{js,ts}'
+import { db } from 'src/lib/db'
+import services from 'src/services/**/*.{js,ts}'
```
All copied from a new redwood project

bozdoz commented on api/src/services/countries/countries.test.ts:
```diff
@@ -0,0 +1,52 @@
+  deleteCountry
+} from './countries'
+
+describe('countries', () => {
```
Guessing redwood created these somehow too

bozdoz commented on package.json:
```diff
@@ -7,13 +7,20 @@
   },
   "engines": {
-    "node": ">=12",
+    "node": ">=14",
```
Updated node version could be a problem for production?

lachlanjc replied:
Should be fine!

bozdoz commented on web/src/components/Chart.js:
```diff
@@ -118,7 +118,7 @@ const Chart = ({
   const countryCounts = groupBy(sortedDailyCounts, 'country.iso')
   // Highest date in benchmark country
-  const maxBenchmarkDate = last(countryCounts[defaultCountry]).date.date
+  const maxBenchmarkDate = last(countryCounts[defaultCountry])?.date.date
```
Fix for issues if database is empty

bozdoz commented:
Docker was a bit tough, but I got it up eventually.

Added a few more countries.  Changed default country to be an environment variable.  Caching one of the queries to prevent re-renders.  Removed redwood-hack directory and seems to work fine.  Also I think I fixed the scraper.

Try it out if you have docker: `docker-compose up --build`

Scraper seems to take about 10min

bozdoz commented:
proof of scraper fix (I think): 🤔 

<img width="1170" alt="image" src="https://user-images.githubusercontent.com/1410985/118070938-3536cd80-b37d-11eb-971a-e477774491d8.png">

bozdoz commented on api/prisma/countries.js:
```diff
@@ -50,24 +43,22 @@ const countries = [
+  {
+    iso: 'cad',
+    worldometersSlug: 'canada',
+    name: 'Canada',
```
closes #6

bozdoz commented on api/prisma/seed.js:
```diff
@@ -0,0 +1,33 @@
+ *
+ * @template T
+ * @param {Array<T>} array
+ * @param {(item: T) => void} callback
```
jsdoc adds types to javascript: https://medium.com/@antonkrinitsyn/jsdoc-generic-types-typescript-db213cf48640

bozdoz commented on api/prisma/seed.js:
```diff
@@ -0,0 +1,33 @@
+async function asyncForEach(array, callback) {
+  const promises = array.map((item) => callback(item))
+
+  return Promise.allSettled(promises)
```
allSettled means that some or all could fail 🤷

bozdoz commented on api/src/functions/graphql.js:
```diff
@@ -3,24 +3,24 @@ import {
-    }),
-  ],
+
+  cors: {
+    origin: '*',
+    credentials: true,
```
cors is just to make docker work; if you want I could make this conditional

bozdoz commented on api/src/functions/scrape.js:
```diff
@@ -1,31 +1,43 @@
-    htmlAttempt = $(`div#${chartId} + script`).html()
+  const hackedChartJSON = htmlAttempt.split('Highcharts.chart(')[1];
+
+  const xAxisCategories = hackedChartJSON.match(/xAxis:.*?{.*?categories\:.*?(\[.*?\])/s)
```
here's where we just get xAxis categories directly

bozdoz commented on api/src/functions/scrape.js:
```diff
@@ -1,31 +1,43 @@
+
+  const xVals = JSON.parse(xAxisCategories[1])
+
+  const seriesData = hackedChartJSON.match(/series:.*?data:.*?(\[.*?\])/s)
```
and this gets series data directly

bozdoz commented on api/src/functions/scrape.js:
```diff
@@ -1,31 +1,43 @@
-    .replace(/^.*?{/, '{')
-    .replace(/\); ?$/, '')
-    .replace(/([A-z]+):/g, '"$1":')
-    .replace(/'/g, '"')
+  const yVals = JSON.parse(seriesData[1])
 
-  const json = JSON.parse(hackedChartJSON)
-  const xVals = json.xAxis.categories
-  const yVals = json.series[0].data
+  console.log("parsed chart json!")
 
   let data = {}
 
   xVals.forEach((val, i) => {
-    // need to manually set year to 2020
-    let day = new Date(val + ", 2020")
+    const day = new Date(val)
```
I assume that's not true anymore

bozdoz commented on api/src/functions/scrape.js:
```diff
@@ -158,7 +163,20 @@ Beginning DailyCount creation for ${day} (or update if it already exists)...`)
+  try {
+    await dbBS()
+    console.log('success')
+    return {
+      statusCode: 200,
+      body: JSON.stringify('Success!')
+    }
```
new format for redwood function handlers

bozdoz commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+
+services:
+  web:
+    build: .
```
builds Dockerfile

bozdoz commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+  web:
+    build: .
+    ports:
+      - 8910:8910
```
binds host port 8910 to container port 8910

bozdoz commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    build: .
+    ports:
+      - 8910:8910
+    command: yarn rw dev web
```
main command to bring up web side

bozdoz commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+      - 8910:8910
+    command: yarn rw dev web
+    restart: always
+    environment:
+      REDWOOD_ENV_DEFAULT_COUNTRY: usa
+      REDWOOD_ENV_DEFAULT_COUNTRY_NAME: United States
```
dev could easily change these if they wanted

bozdoz commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    restart: always
+    volumes:
+      # keeps database after container stops
+      - db:/app/api/prisma
```
this is a named volume which docker keeps internally

bozdoz commented on netlify.toml:
```diff
@@ -1,5 +1,5 @@
 [build]
-  command = "cp redwood-hack/api-dbInstance.js node_modules/@redwoodjs/api/dist/dbInstance.js && yarn rw db up --no-db-client && yarn rw build"
+  command = "yarn rw prisma migrate dev && yarn rw build"
```
not sure

bozdoz commented on web/src/components/CountriesCell/CountriesCell.js:
```diff
@@ -2,6 +2,19 @@ import { find, orderBy, filter, concat } from 'lodash'
+export const beforeQuery = () => {
+  return {
+    // never change query with new variables
+    variables,
```
the query (below) never uses variables, so this prevents it from fetching what it thinks is a new query

bozdoz commented on web/src/components/CountriesCell/CountriesCell.js:
```diff
@@ -2,6 +2,19 @@ import { find, orderBy, filter, concat } from 'lodash'
+    variables,
+    fetchPolicy: 'cache-and-network',
+    // countries don't change mid-session
+    nextFetchPolicy: 'no-cache'
```
meant for this to be cache-only, but maybe doesn't matter

bozdoz replied:
cache-only breaks. Maybe should revert to cache-first (default)
